README.md
# quRan <img src="tools/logo.png" align="right"/>

## An R package for the full text of the Qur'an

This package contains the complete text of the Qur'an in Arabic (with and without vowels) and in English (the Yusuf Ali and Saheeh International translations).

The text for each volume or translation was sourced from [Tanzil](http://tanzil.net/docs/download) and [Al Quran Cloud](https://alquran.cloud/).

Each volume is in a data frame with a row for each ayah, along with 18 columns of detailed metadata (including ayah number, surah number, surah name in Arabic and English, juz', hizb, and so on), ready for text analysis. 

The package contains:

- `quran_ar`: The Qur'an in Arabic (with vowels)
- `quran_ar_min`: The Qur'an in Arabic (without vowels)
- `quran_en_yusufali`: The Yusuf Ali English translation of the Qur'an
- `quran_en_sahih`: The Saheeh International English translation of the Qur'an

Surah titles in Arabic, transliterated English, and translated English are all factors to preserve the correct order.


## Installation

This package is not released on CRAN yet. You can install the development version from Github:

```r
library(remotes)
install_github("andrewheiss/quRan")
library(quRan)
```


## Usage

For some ideas on getting started with analyzing these texts, see [*Text Mining with R* by Julia Silge and David Robinson](https://www.tidytextmining.com/), or Julia Silge's [blog post on sentiment analysis of Jane Austen's novels](https://juliasilge.com/blog/if-i-loved-nlp-less/). For help within R, try `?quran_ar` or similar for getting started with the datasets.

Here's how to do some basic data manipulation with the data:

``` r
library(dplyr)
library(quRan)

quran_ar %>% 
  group_by(surah_title_en) %>% 
  summarize(total_ayat = n())
#> # A tibble: 114 x 2
#>    surah_title_en total_ayat
#>    <fct>               <int>
#>  1 Al-Faatiha              7
#>  2 Al-Baqara             286
#>  3 Aal-i-Imraan          200
#>  4 An-Nisaa              176
#>  5 Al-Maaida             120
#>  6 Al-An'aam             165
#>  7 Al-A'raaf             206
#>  8 Al-Anfaal              75
#>  9 At-Tawba              129
#> 10 Yunus                 109
#> # ... with 104 more rows

quran_ar %>% 
  filter(surah_title_en == "Al-Faatiha") %>% 
  select(surah, ayah, text)
#> # A tibble: 7 x 3
#>   surah  ayah text                                                
#>   <int> <int> <chr>                                               
#> 1     1     1 بِسْمِ اللَّهِ الرَّحْمَٰنِ الرَّحِيمِ                              
#> 2     1     2 الْحَمْدُ لِلَّهِ رَبِّ الْعَالَمِينَ                               
#> 3     1     3 الرَّحْمَٰنِ الرَّحِيمِ                                       
#> 4     1     4 مَالِكِ يَوْمِ الدِّينِ                                      
#> 5     1     5 إِيَّاكَ نَعْبُدُ وَإِيَّاكَ نَسْتَعِينُ                              
#> 6     1     6 اهْدِنَا الصِّرَاطَ الْمُسْتَقِيمَ                               
#> 7     1     7 صِرَاطَ الَّذِينَ أَنْعَمْتَ عَلَيْهِمْ غَيْرِ الْمَغْضُوبِ عَلَيْهِمْ وَلَا الضَّالِّينَ
```


## Code of Conduct

This project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.


## Credits

The text and translations are all open source or licensed under Creative Commons by [Tanzil](http://tanzil.net/docs/download) and [Al Quran Cloud](https://alquran.cloud/). Complete details are found in the [LICENSE](/LICENSE) file.

Fatihah background in hex icon from [seeklogo](https://seeklogo.com/vector-logo/183552/sureh-fatiha). 


NEWS.md
# quRan 0.1.0

- Initial release of full text of the Qur'an in Arabic and English


tests/testthat.R
library(testthat)
library(quRan)

test_check("quRan")


tests/testthat/test_quran.R
# tests for the Qur'an data

context("Tidy dataframe with Qur'an text")

suppressPackageStartupMessages(library(dplyr))

test_that("factor order is correct", {
  expect_equal(levels(quran_ar$surah_title_en)[1], "Al-Faatiha")
  expect_equal(levels(quran_ar$surah_title_en_trans)[1], "The Opening")
  expect_equal(levels(quran_ar$surah_title_ar)[1], "الفاتحة")

  expect_equal(levels(quran_ar$surah_title_en)[114], "An-Naas")
  expect_equal(levels(quran_ar$surah_title_en_trans)[114], "Mankind")
  expect_equal(levels(quran_ar$surah_title_ar)[114], "الناس")
})

test_that("tidy frame for the Qu'ran in Arabic is right", {
  # Does this have the right number of lines?
  expect_equal(nrow(quran_ar), 6236)
  expect_equal(ncol(quran_ar), 18)

  # Are the lines spread across each volume correctly?
  surahs_test <- quran_ar %>%
    group_by(surah_title_en) %>%
    summarise(total_ayahs = n())
  expect_equal(nrow(surahs_test), 114)
  expect_equal(ncol(surahs_test), 2)
})


data-raw/clean_data.R
library(dplyr)
library(tidyr)
library(stringr)
library(purrr)
library(jsonlite)


# Download raw data from alquran.cloud ------------------------------------

# Arabic, all vowels
download.file("http://api.alquran.cloud/quran/quran-simple",
              "./data-raw/quran-simple.json")
quran_ar_raw <- read_json("./data-raw/quran-simple.json")

# Arabic, no vowels
download.file("http://api.alquran.cloud/quran/quran-simple-clean",
              "./data-raw/quran-simple-clean.json")
quran_clean_ar_raw <- read_json("./data-raw/quran-simple-clean.json")

# English, Yusuf Ali
download.file("http://api.alquran.cloud/quran/en.yusufali",
              "./data-raw/en.yusufali.json")
quran_en_yusufali_raw <- read_json("./data-raw/en.yusufali.json")

# English, Saheeh International
download.file("http://api.alquran.cloud/quran/en.sahih",
              "./data-raw/en.sahih.json")
quran_en_sahih_raw <- read_json("./data-raw/en.sahih.json")



# Load and clean raw data -------------------------------------------------

# There are 15 sajdas, and each has additional metadata (like whether it is is
# recommended or obligatory) that is in a nested list. If we just use
# as_data_frame(), ayahs with sajdas turn into a three-row data frame that can't
# be bound with the sajda-free ayahs. So, we have to treat sajda ayahs a little
# differently, flattening them and renaming some of the columns
extract_ayahs <- function(x) {
  n_sajda <- length(x$sajda)

  if (n_sajda == 1) {
    as_data_frame(x)
  } else {
    as_data_frame(purrr::flatten(x)) %>%
      mutate(sajda = TRUE) %>%
      rename(sajda_id = id, sajda_recommended = recommended,
             sajda_obligatory = obligatory)
  }
}

# Extract all the surah and ayah details
quran_ar <- data_frame(surah_id = 1:114) %>%
  mutate(surah_details =
           map(surah_id, ~ as_data_frame(quran_ar_raw$data$surahs[[.x]]))) %>%
  unnest(surah_details) %>%
  mutate(ayahs = ayahs %>% map(~ extract_ayahs(.x))) %>%
  unnest(ayahs) %>%
  mutate(ayah_title = paste0(surah_id, ":", numberInSurah),
         surah = surah_id) %>%
  select(surah_id, ayah_id = number1,
         surah_title_ar = name, surah_title_en = englishName,
         surah_title_en_trans = englishNameTranslation,
         revelation_type = revelationType,
         text, surah, ayah = numberInSurah, ayah_title,
         juz, manzil, page, hizb_quarter = hizbQuarter,
         sajda, sajda_id, sajda_recommended, sajda_obligatory) %>%
  mutate(surah_title_ar = str_replace(surah_title_ar, "سورة ", "")) %>%
  # There are two "The Dawns" so make 113 Al-Falaq be "The Daybreak"
  mutate(surah_title_en_trans = ifelse(surah_id == 113, "The Daybreak", surah_title_en_trans)) %>%
  mutate_at(vars(surah_title_ar, surah_title_en, surah_title_en_trans),
            funs(forcats::fct_inorder))

quran_ar_min <- data_frame(surah_id = 1:114) %>%
  mutate(surah_details =
           map(surah_id, ~ as_data_frame(quran_clean_ar_raw$data$surahs[[.x]]))) %>%
  unnest(surah_details) %>%
  mutate(ayahs = ayahs %>% map(~ extract_ayahs(.x))) %>%
  unnest(ayahs) %>%
  mutate(ayah_title = paste0(surah_id, ":", numberInSurah),
         surah = surah_id) %>%
  select(surah_id, ayah_id = number1,
         surah_title_ar = name, surah_title_en = englishName,
         surah_title_en_trans = englishNameTranslation,
         revelation_type = revelationType,
         text, surah, ayah = numberInSurah, ayah_title,
         juz, manzil, page, hizb_quarter = hizbQuarter,
         sajda, sajda_id, sajda_recommended, sajda_obligatory) %>%
  mutate(surah_title_ar = str_replace(surah_title_ar, "سورة ", "")) %>%
  # There are two "The Dawns" so make 113 Al-Falaq be "The Daybreak"
  mutate(surah_title_en_trans = ifelse(surah_id == 113, "The Daybreak", surah_title_en_trans)) %>%
  mutate_at(vars(surah_title_ar, surah_title_en, surah_title_en_trans),
            funs(forcats::fct_inorder))

quran_en_yusufali <- data_frame(surah_id = 1:114) %>%
  mutate(surah_details =
           map(surah_id, ~ as_data_frame(quran_en_yusufali_raw$data$surahs[[.x]]))) %>%
  unnest(surah_details) %>%
  mutate(ayahs = ayahs %>% map(~ extract_ayahs(.x))) %>%
  unnest(ayahs) %>%
  mutate(ayah_title = paste0(surah_id, ":", numberInSurah),
         surah = surah_id) %>%
  select(surah_id, ayah_id = number1,
         surah_title_ar = name, surah_title_en = englishName,
         surah_title_en_trans = englishNameTranslation,
         revelation_type = revelationType,
         text, surah, ayah = numberInSurah, ayah_title,
         juz, manzil, page, hizb_quarter = hizbQuarter,
         sajda, sajda_id, sajda_recommended, sajda_obligatory) %>%
  mutate(surah_title_ar = str_replace(surah_title_ar, "سورة ", "")) %>%
  # There are two "The Dawns" so make 113 Al-Falaq be "The Daybreak"
  mutate(surah_title_en_trans = ifelse(surah_id == 113, "The Daybreak", surah_title_en_trans)) %>%
  mutate_at(vars(surah_title_ar, surah_title_en, surah_title_en_trans),
            funs(forcats::fct_inorder))

quran_en_sahih <- data_frame(surah_id = 1:114) %>%
  mutate(surah_details =
           map(surah_id, ~ as_data_frame(quran_en_sahih_raw$data$surahs[[.x]]))) %>%
  unnest(surah_details) %>%
  mutate(ayahs = ayahs %>% map(~ extract_ayahs(.x))) %>%
  unnest(ayahs) %>%
  mutate(ayah_title = paste0(surah_id, ":", numberInSurah),
         surah = surah_id) %>%
  select(surah_id, ayah_id = number1,
         surah_title_ar = name, surah_title_en = englishName,
         surah_title_en_trans = englishNameTranslation,
         revelation_type = revelationType,
         text, surah, ayah = numberInSurah, ayah_title,
         juz, manzil, page, hizb_quarter = hizbQuarter,
         sajda, sajda_id, sajda_recommended, sajda_obligatory) %>%
  mutate(surah_title_ar = str_replace(surah_title_ar, "سورة ", "")) %>%
  # There are two "The Dawns" so make 113 Al-Falaq be "The Daybreak"
  mutate(surah_title_en_trans = ifelse(surah_id == 113, "The Daybreak", surah_title_en_trans)) %>%
  mutate_at(vars(surah_title_ar, surah_title_en, surah_title_en_trans),
            funs(forcats::fct_inorder))


# Add to package ----------------------------------------------------------

devtools::use_data(quran_ar, overwrite = TRUE)
devtools::use_data(quran_ar_min, overwrite = TRUE)
devtools::use_data(quran_en_yusufali, overwrite = TRUE)
devtools::use_data(quran_en_sahih, overwrite = TRUE)


R/quran.R
#' The Qur'an
#'
#' This package contains the complete text of the Qur'an in Arabic (with and
#' without vowels) and in English (the Yusuf Ali and Saheeh International
#' translations), formatted to be convenient for text analysis.
#' @docType package
#' @name quRan
#' @aliases quRan quRan-package
NULL

#' The text of the Qur'an (Arabic, all vowels)
#'
#' A dataset containing the Arabic text of the Qur'an, with vowels.
#'
#' Because Surahs 89 and 113 are both translated as "The Dawn," Surah 113 has
#' been retitled "The Daybreak"
#'
#' @source Tanzil (\url{http://tanzil.net/docs/download}) and Al Quran Cloud (\url{https://alquran.cloud/})
#' @format A data frame with 6236 rows and 18 columns:
#' \describe{
#'   \item{\code{surah_id}}{Unique id number for the surah}
#'   \item{\code{ayah_id}}{Unique id number for the ayah}
#'   \item{\code{surah_title_ar}}{Name of the surah (Arabic)}
#'   \item{\code{surah_title_en}}{Name of the surah (English; transliterated)}
#'   \item{\code{surah_title_en_trans}}{Name of the surah (English; translated)}
#'   \item{\code{revelation_type}}{Type of the surah (Meccan or Medinan)}
#'   \item{\code{text}}{Ayah text}
#'   \item{\code{surah}}{Surah}
#'   \item{\code{ayah}}{Ayah}
#'   \item{\code{ayah_title}}{Combined surah and ayah (e.g. 2:242)}
#'   \item{\code{juz}}{Juz'}
#'   \item{\code{manzil}}{Manzil}
#'   \item{\code{page}}{Page number}
#'   \item{\code{hizb_quarter}}{Maqra}
#'   \item{\code{sajda}}{Binary indicator for presence of a sajdah}
#'   \item{\code{sajda_id}}{Unique id number for the sajdah}
#'   \item{\code{sajda_recommended}}{Binary indicator for whether a sajdah is recommended}
#'   \item{\code{sajda_obligatory}}{Binary indicator for whether a sajdah is obligatory}
#' }
"quran_ar"

#' The text of the Qur'an (Arabic, no vowels)
#'
#' A dataset containing the Arabic text of the Qur'an, without vowels.
#'
#' Because Surahs 89 and 113 are both translated as "The Dawn," Surah 113 has
#' been retitled "The Daybreak"
#'
#' @source Tanzil (\url{http://tanzil.net/docs/download}) and Al Quran Cloud (\url{https://alquran.cloud/})
#' @format A data frame with 6236 rows and 18 columns:
#' \describe{
#'   \item{\code{surah_id}}{Unique id number for the surah}
#'   \item{\code{ayah_id}}{Unique id number for the ayah}
#'   \item{\code{surah_title_ar}}{Name of the surah (Arabic)}
#'   \item{\code{surah_title_en}}{Name of the surah (English; transliterated)}
#'   \item{\code{surah_title_en_trans}}{Name of the surah (English; translated)}
#'   \item{\code{revelation_type}}{Type of the surah (Meccan or Medinan)}
#'   \item{\code{text}}{Ayah text}
#'   \item{\code{surah}}{Surah}
#'   \item{\code{ayah}}{Ayah}
#'   \item{\code{ayah_title}}{Combined surah and ayah (e.g. 2:242)}
#'   \item{\code{juz}}{Juz'}
#'   \item{\code{manzil}}{Manzil}
#'   \item{\code{page}}{Page number}
#'   \item{\code{hizb_quarter}}{Maqra}
#'   \item{\code{sajda}}{Binary indicator for presence of a sajdah}
#'   \item{\code{sajda_id}}{Unique id number for the sajdah}
#'   \item{\code{sajda_recommended}}{Binary indicator for whether a sajdah is recommended}
#'   \item{\code{sajda_obligatory}}{Binary indicator for whether a sajdah is obligatory}
#' }
"quran_ar_min"

#' The Yusuf Ali translation of the Qur'an (English)
#'
#' A dataset containing the English text of the Yusuf Ali translation of the Qur'an.
#'
#' Because Surahs 89 and 113 are both translated as "The Dawn," Surah 113 has
#' been retitled "The Daybreak"
#'
#' @source Tanzil (\url{http://tanzil.net/docs/download}) and Al Quran Cloud (\url{https://alquran.cloud/})
#' @format A data frame with 6236 rows and 18 columns:
#' \describe{
#'   \item{\code{surah_id}}{Unique id number for the surah}
#'   \item{\code{ayah_id}}{Unique id number for the ayah}
#'   \item{\code{surah_title_ar}}{Name of the surah (Arabic)}
#'   \item{\code{surah_title_en}}{Name of the surah (English; transliterated)}
#'   \item{\code{surah_title_en_trans}}{Name of the surah (English; translated)}
#'   \item{\code{revelation_type}}{Type of the surah (Meccan or Medinan)}
#'   \item{\code{text}}{Ayah text}
#'   \item{\code{surah}}{Surah}
#'   \item{\code{ayah}}{Ayah}
#'   \item{\code{ayah_title}}{Combined surah and ayah (e.g. 2:242)}
#'   \item{\code{juz}}{Juz'}
#'   \item{\code{manzil}}{Manzil}
#'   \item{\code{page}}{Page number}
#'   \item{\code{hizb_quarter}}{Maqra}
#'   \item{\code{sajda}}{Binary indicator for presence of a sajdah}
#'   \item{\code{sajda_id}}{Unique id number for the sajdah}
#'   \item{\code{sajda_recommended}}{Binary indicator for whether a sajdah is recommended}
#'   \item{\code{sajda_obligatory}}{Binary indicator for whether a sajdah is obligatory}
#' }
"quran_en_yusufali"

#' The Saheeh International translation of the Qur'an (English)
#'
#' A dataset containing the English text of the Saheeh International translation of the Qur'an.
#'
#' Because Surahs 89 and 113 are both translated as "The Dawn," Surah 113 has
#' been retitled "The Daybreak"
#'
#' @source Tanzil (\url{http://tanzil.net/docs/download}) and Al Quran Cloud (\url{https://alquran.cloud/})
#' @format A data frame with 6236 rows and 18 columns:
#' \describe{
#'   \item{\code{surah_id}}{Unique id number for the surah}
#'   \item{\code{ayah_id}}{Unique id number for the ayah}
#'   \item{\code{surah_title_ar}}{Name of the surah (Arabic)}
#'   \item{\code{surah_title_en}}{Name of the surah (English; transliterated)}
#'   \item{\code{surah_title_en_trans}}{Name of the surah (English; translated)}
#'   \item{\code{revelation_type}}{Type of the surah (Meccan or Medinan)}
#'   \item{\code{text}}{Ayah text}
#'   \item{\code{surah}}{Surah}
#'   \item{\code{ayah}}{Ayah}
#'   \item{\code{ayah_title}}{Combined surah and ayah (e.g. 2:242)}
#'   \item{\code{juz}}{Juz'}
#'   \item{\code{manzil}}{Manzil}
#'   \item{\code{page}}{Page number}
#'   \item{\code{hizb_quarter}}{Maqra}
#'   \item{\code{sajda}}{Binary indicator for presence of a sajdah}
#'   \item{\code{sajda_id}}{Unique id number for the sajdah}
#'   \item{\code{sajda_recommended}}{Binary indicator for whether a sajdah is recommended}
#'   \item{\code{sajda_obligatory}}{Binary indicator for whether a sajdah is obligatory}
#' }
"quran_en_sahih"
